splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/CrossJoinOperation.java

```diff
@@ -0,0 +1,198 @@
+                              boolean oneRowRightSide,
+                              boolean notExistsRightSide,
+                              boolean rightFromSSQ,
+                              double optimizerEstimatedRowCount,
+                              double optimizerEstimatedCost,
+                              String userSuppliedOptimizerOverrides) throws
+            StandardException{
+        super(leftResultSet,leftNumCols,rightResultSet,rightNumCols,
+                activation,restriction,resultSetNumber,oneRowRightSide,notExistsRightSide, rightFromSSQ,
+                optimizerEstimatedRowCount,optimizerEstimatedCost,userSuppliedOptimizerOverrides);
+        this.leftHashKeyItem=leftHashKeyItem;
+        this.rightHashKeyItem=rightHashKeyItem;
+        this.sequenceId = Bytes.toLong(operationInformation.getUUIDGenerator().nextBytes());
+        init();
+    }
+
+    @Override
+    public void readExternal(ObjectInput in) throws IOException,
+            ClassNotFoundException{
+        super.readExternal(in);
+        sequenceId = in.readLong();
+        leftHashKeyItem=in.readInt();
+        rightHashKeyItem=in.readInt();
+    }
+
+    public long getSequenceId() {
+        return sequenceId;
+    }
+
+    @Override
+    public void writeExternal(ObjectOutput out) throws IOException{
+        super.writeExternal(out);
+        out.writeLong(sequenceId);
+        out.writeInt(leftHashKeyItem);
+        out.writeInt(rightHashKeyItem);
+    }
+
+    @Override
+    public void init(SpliceOperationContext context) throws StandardException, IOException {
+        super.init(context);
+        leftHashKeys = generateHashKeys(leftHashKeyItem);
+        rightHashKeys = generateHashKeys(rightHashKeyItem);
+    }
+
+    @Override
+    public SpliceOperation getLeftOperation(){
+        return leftResultSet;
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public DataSet<ExecRow> getDataSet(DataSetProcessor dsp) throws StandardException {
+        if (!isOpen)
+            throw new IllegalStateException("Operation is not open");
+
+        OperationContext operationContext = dsp.createOperationContext(this);
+        DataSet<ExecRow> leftDataSet = leftResultSet.getDataSet(dsp);
+        DataSet<ExecRow> rightDataSet = rightResultSet.getDataSet(dsp);
+
+//        operationContext.pushScope();
+        leftDataSet = leftDataSet.map(new CountJoinedLeftFunction(operationContext));
+
+        DataSet<ExecRow> result;
+
+        if (dsp.getType().equals(DataSetProcessor.Type.SPARK)) {
+            result = leftDataSet.crossJoin(operationContext, rightDataSet);
+            if (restriction != null) {
+                result = result.filter(new JoinRestrictionPredicateFunction(operationContext));
+            }
+        } else {
+            LOG.warn("Cross join supposed to be run with Spark only, using BroadcastJoin now");
+            if (isOuterJoin) { // Outer Join with and without restriction
```

Comment:
I don't think this should be a case here, a cross join shouldn't be an outer join

Reply:
Yes, it should be rule out in the optimizer. I can try to remove it and test it again.